dm/pkg/router/router_test.go

```diff
@@ -0,0 +1,387 @@
+	for _, rule := range rules {
+		err := r.AddRule(rule)
+		c.Assert(err, Equals, nil) // successfully insert
+	}
+}
+
+func (s *testRouteSuite) TestSchemaRoute(c *C) {
+	rules := []*TableRule{
+		{
+			SchemaPattern: "test1",
+			TargetSchema:  "dtest1",
+		},
+		{
+			SchemaPattern: "gtest*",
+			TargetSchema:  "dtest",
+		},
+	}
+	oldRouter, err := oldrouter.NewTableRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	newRouter, err := NewRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	inputTables := []Table{
+		{
+			Schema: "test1", // match rule 1
+			Name:   "table1",
+		},
+		{
+			Schema: "gtesttest", // match rule 2
+			Name:   "atable",
+		},
+		{
+			Schema: "ptest", // match neither
+			Name:   "atableg",
+		},
+	}
+	expectedResult := []Table{
+		{
+			Schema: "dtest1",
+			Name:   "table1",
+		},
+		{
+			Schema: "dtest",
+			Name:   "atable",
+		},
+		{
+			Schema: "ptest",
+			Name:   "atableg",
+		},
+	}
+	for idx := range inputTables {
+		schema, table := inputTables[idx].Schema, inputTables[idx].Name
+		expSchema, expTable := expectedResult[idx].Schema, expectedResult[idx].Name
+		oldSchema, oldTable, err := oldRouter.Route(schema, table)
+		c.Assert(err, Equals, nil)
+		newSchema, newTable, err := newRouter.Route(schema, table)
+		c.Assert(err, Equals, nil)
+		c.Assert(oldSchema, Equals, expSchema)
+		c.Assert(oldTable, Equals, expTable)
+		c.Assert(newSchema, Equals, expSchema)
+		c.Assert(newTable, Equals, expTable)
+	}
+}
+
+func (s *testRouteSuite) TestTableRoute(c *C) {
+	rules := []*TableRule{
+		{
+			SchemaPattern: "test1",
+			TablePattern:  "table1",
+			TargetSchema:  "dtest1",
+			TargetTable:   "dtable1",
+		},
+		{
+			SchemaPattern: "test*",
+			TablePattern:  "table2",
+			TargetSchema:  "dtest2",
+			TargetTable:   "dtable2",
+		},
+		{
+			SchemaPattern: "test3",
+			TablePattern:  "table*",
+			TargetSchema:  "dtest3",
+			TargetTable:   "dtable3",
+		},
+	}
+	inputTables := []*Table{}
+	expTables := []*Table{}
+	for i := 1; i <= 3; i++ {
+		inputTables = append(inputTables, &Table{
+			Schema: fmt.Sprintf("test%d", i),
+			Name:   fmt.Sprintf("table%d", i),
+		})
+		expTables = append(expTables, &Table{
+			Schema: fmt.Sprintf("dtest%d", i),
+			Name:   fmt.Sprintf("dtable%d", i),
+		})
+	}
+	oldRouter, err := oldrouter.NewTableRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	newRouter, err := NewRouter(true, rules)
+	c.Assert(err, Equals, nil)
+	for i := range inputTables {
+		schema, table := inputTables[i].Schema, inputTables[i].Name
+		expSchema, expTable := expTables[i].Schema, expTables[i].Name
+		oldSch, oldTbl, _ := oldRouter.Route(schema, table)
+		newSch, newTbl, _ := newRouter.Route(schema, table)
+		c.Assert(newSch, Equals, expSchema)
+		c.Assert(newTbl, Equals, expTable)
+		c.Assert(oldSch, Equals, expSchema)
+		c.Assert(oldTbl, Equals, expTable)
+	}
+}
+
+func (s *testRouteSuite) TestRegExprRoute(c *C) {
+	rules := []*TableRule{
+		{
+			SchemaPattern: "~test.[0-9]+",
```

Comment:
can you give a meaningful string, it's very hard to check the result :cry:

Reply:
Would you explain more about the hard reading part?

Reply:
I need to look back to `~test3(gd|ct)?8917*` to check `test3gd8917777` is matched as expected (does it have same four digit?) and check other regexp is not matced as expected (like `~test.[0-9]+`), a lot of back and forth jumping.

if they can be `~test1[0-9]*`, `~test2[0-9]*`, `~test[a-z][0-9]*` or other pattern it would be easier to remember.

Reply:
Yes, it's a bit complicated. Actually I'm trying to build some similar but exclusive patterns and see if they conflicts with the rule that only one routing rule is matched. It's perhaps unnecessarily complex😂. I'll fix it later...